### ISSUE TITLE
Option to hide collision shapes of instanced scenes

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#54]
**Describe the project you are working on:**
A complex game with lots of object types.

**Describe how this feature / enhancement will help your project:**
I have one or more collision shapes on almost every object in my game. These objects are of course separate scenes, instanced in maps. Seeing these shapes all over the scene hurts readability and is... annoying >_>
![image](https://user-images.githubusercontent.com/2223172/64609956-6dc71c80-d3ce-11e9-940a-8bfdd93c6e3f.png)
(ignore placeholder sprites)

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
Well, this could be editor option. After enabling, if you instance a scene with collision shapes, the shapes would be hidden (similar to how they don't show while playing).
![image](https://user-images.githubusercontent.com/2223172/64610208-0d84aa80-d3cf-11e9-8451-3660ff798f56.png)
Much more pleasant to look at, isn't it?

They would still be visible in the original scene.

EDIT:
I just noticed the View menu in the editor. Perfect place for this option.

**Describe implementation detail for your proposal (in code), if possible:**
For 2D, I guess it's about adding another case here
https://github.com/godotengine/godot/blob/24e1039eb6fe32115e8d1a62a84965e9be19a2ed/scene/2d/collision_shape_2d.cpp#L104
which would check whether the shape is part of instanced scene or its own.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
You can manually turn off the visibility of the shapes, but that state is needlessly saved in the scene and also you need to toggle it each time you want to edit the shape. It's just bothersome.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's a little QoL thing. Not sure if possible with a reasonable plugin (reasonable i.e. not adding any additional steps to simple auto-hiding).

**Comment by madmiraal** [comment]
I don't think this is a physics proposal. I think it's an editor proposal.

That been said, I think this is unnecessary, because it can achieved by simply changing the order of the `Sprite` and `CollisionShape2D` nodes.

If the `CollisionShape2D` node is placed before the `Sprite` node, the `Sprite` is drawn on top of the `CollisionShape2D` node, and visa versa:
![CollisionShape1](https://user-images.githubusercontent.com/9253928/80352512-e7335280-886b-11ea-981e-d221cfad03a4.png)
![CollisionShape2](https://user-images.githubusercontent.com/9253928/80352522-eb5f7000-886b-11ea-86a1-9802cbd40595.png)

**Comment by KoBeWi** [comment]
> If the CollisionShape2D node is placed before the Sprite node, the Sprite is drawn on top of the CollisionShape2D node, and visa versa:

Irregular Sprites will get a very visible background though. Also I sometimes make additional shapes for detectors and it's impossible to cover them.

**Comment by nhold** [comment]
@madmiraal That won't work if the parent node is the sprite or the collision shape is much bigger than the sprite.

**Comment by KoBeWi** [comment]
Putting sprites on the top turned out to be plausible solution. It just requires reordering scenes and bigger colliders are rare enough that hiding them manually is ok. So I don't think this proposal is worth it to be implemented.

**Comment by Zireael07** [comment]
@KoBeWi: Sprites on top is only a solution for 2D. For 3D, the problem remains. Please reopen?

**Comment by KoBeWi** [comment]
I didn't even consider 3D for this feature, but ok.

**Comment by madmiraal** [comment]
> Sprites on top is only a solution for 2D. For 3D, the problem remains.

Unless I'm missing something, the same plausible solution applies to 3D too i.e. put the `MeshInstance` after the `CollisionShape`.

**Comment by nhold** [comment]
> 
> 
> Putting sprites on the top turned out to be plausible solution. It just requires reordering scenes and bigger colliders are rare enough that hiding them manually is ok. So I don't think this proposal is worth it to be implemented.

Bigger collision shapes are rare? Vision radius/arc. Ability collision shapes/'hidden' trap collision shapes. There are many ways to have a collision shape that pollutes the editor (Even if it is possible to re-order scenes). I still think this is a good proposal for 2D.

If we add (Or already exists, I don't know) a hook for when a scene is instanced\loaded in editor we could potentially have our own toggle.

**Comment by Calinou** [comment]
If https://github.com/godotengine/godot/pull/40123 was implemented for 2D, it would likely address this proposal since you could use a very low opacity.

**Comment by GoldenSkullArt** [comment]
None of the suggested "solutions" are really a solution if you have a lot of objects in the scene, to just toggle the visibility of all scene-object's colliders in the parent scene would help sooo much, then I don't have to manually toggle them all on and off, when debugging or when not debugging.

**Comment by me2beats** [comment]
yep this is kinda annoying
isn't it somewhere in theme settings (if so I could just set its alpha). or the color is hardcoded?
And I guess it is drawn using a draw_.. function
![изображение](https://user-images.githubusercontent.com/16458555/168183157-f61c7549-e1f6-4ad5-aa48-e438bdd5fc88.png)

**Comment by Calinou** [comment]
> yep this is kinda annoying isn't it somewhere in theme settings (if so I could just set its alpha). or the color is hardcoded? And I guess it is drawn using a draw_.. function 

You should be able to change the **Self Modulate** property on the CollisionShape2D.

**Comment by me2beats** [comment]
having it in editor theme for being able to tweak it globally + an option to hide it completely or only for scene instances, still looks more convenient to me

**Comment by Pacheco95** [comment]
Would you accept a PR for this (at least for 2D)?

**Comment by starblinky** [comment]
As a new user, this is something I would really appreciate as a feature. Sometimes I want to turn off collider visibility just to see the aesthetic of the scene. Or because the visual clutter is making things hard for working in.

Something like View > [Toggle Collider Visibility], or something to toggle visibility off/on

**Comment by Calinou** [comment]
> Would you accept a PR for this (at least for 2D)?

Yes, but we try to keep feature parity between 2D and 3D as much as possible. So we'd prefer if the PR addressed this for both 2D and 3D at the same time.

That said, if you don't know how to implement this in 3D, feel free to open a PR anyway and someone else may be able to assist.

**Comment by dgale763** [comment]
Not a perfect solution - but for anybody just wanting to hide some collision shapes in their project, I believe you can just set the "modulate" property of the collision shape to have an alpha of zero.
